File: taiwan_stock_app/frontend/lib/models/trading.dart
/// 模擬交易模型

class VirtualAccount {
  final int id;
  final int userId;
  final double initialBalance;
  final double cashBalance;
  final double totalValue;
  final double totalProfitLoss;
  final double totalProfitLossPercent;
  final DateTime? createdAt;

  VirtualAccount({
    required this.id,
    required this.userId,
    required this.initialBalance,
    required this.cashBalance,
    required this.totalValue,
    required this.totalProfitLoss,
    required this.totalProfitLossPercent,
    this.createdAt,
  });

  factory VirtualAccount.fromJson(Map<String, dynamic> json) {
    return VirtualAccount(
      id: json['id'],
      userId: json['user_id'],
      initialBalance: (json['initial_balance'] ?? 0).toDouble(),
      cashBalance: (json['cash_balance'] ?? 0).toDouble(),
      totalValue: (json['total_value'] ?? 0).toDouble(),
      totalProfitLoss: (json['total_profit_loss'] ?? 0).toDouble(),
      totalProfitLossPercent: (json['total_profit_loss_percent'] ?? 0).toDouble(),
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'])
          : null,
    );
  }
}

class VirtualPosition {
  final int id;
  final String stockId;
  final String? stockName;
  final int quantity;
  final double avgCost;
  final double? currentPrice;
  final double? marketValue;
  final double unrealizedPnl;
  final double unrealizedPnlPercent;

  VirtualPosition({
    required this.id,
    required this.stockId,
    this.stockName,
    required this.quantity,
    required this.avgCost,
    this.currentPrice,
    this.marketValue,
    required this.unrealizedPnl,
    required this.unrealizedPnlPercent,
  });

  factory VirtualPosition.fromJson(Map<String, dynamic> json) {
    return VirtualPosition(
      id: json['id'],
      stockId: json['stock_id'],
      stockName: json['stock_name'],
      quantity: json['quantity'] ?? 0,
      avgCost: (json['avg_cost'] ?? 0).toDouble(),
      currentPrice: json['current_price']?.toDouble(),
      marketValue: json['market_value']?.toDouble(),
      unrealizedPnl: (json['unrealized_pnl'] ?? 0).toDouble(),
      unrealizedPnlPercent: (json['unrealized_pnl_percent'] ?? 0).toDouble(),
    );
  }

  bool get isProfit => unrealizedPnl > 0;
  bool get isLoss => unrealizedPnl < 0;
}

class VirtualOrder {
  final int id;
  final String stockId;
  final String? stockName;
  final String orderType;
  final int quantity;
  final double price;
  final int filledQuantity;
  final double? filledPrice;
  final String status;
  final DateTime? createdAt;
  final DateTime? filledAt;

  VirtualOrder({
    required this.id,
    required this.stockId,
    this.stockName,
    required this.orderType,
    required this.quantity,
    required this.price,
    required this.filledQuantity,
    this.filledPrice,
    required this.status,
    this.createdAt,
    this.filledAt,
  });

  factory VirtualOrder.fromJson(Map<String, dynamic> json) {
    return VirtualOrder(
      id: json['id'],
      stockId: json['stock_id'],
      stockName: json['stock_name'],
      orderType: json['order_type'],
      quantity: json['quantity'] ?? 0,
      price: (json['price'] ?? 0).toDouble(),
      filledQuantity: json['filled_quantity'] ?? 0,
      filledPrice: json['filled_price']?.toDouble(),
      status: json['status'] ?? 'PENDING',
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'])
          : null,
      filledAt: json['filled_at'] != null
          ? DateTime.tryParse(json['filled_at'])
          : null,
    );
  }

  bool get isBuy => orderType == 'BUY';
  bool get isSell => orderType == 'SELL';
  bool get isFilled => status == 'FILLED';
  bool get isPending => status == 'PENDING';
}

class AccountSummary {
  final VirtualAccount account;
  final List<VirtualPosition> positions;
  final List<VirtualOrder> recentOrders;

  AccountSummary({
    required this.account,
    required this.positions,
    required this.recentOrders,
  });

  factory AccountSummary.fromJson(Map<String, dynamic> json) {
    return AccountSummary(
      account: VirtualAccount.fromJson(json['account']),
      positions: (json['positions'] as List?)
              ?.map((e) => VirtualPosition.fromJson(e))
              .toList() ??
          [],
      recentOrders: (json['recent_orders'] as List?)
              ?.map((e) => VirtualOrder.fromJson(e))
              .toList() ??
          [],
    );
  }
}

File: taiwan_stock_app/backend/app/models/trading.py
"""
模擬交易模型
"""
from sqlalchemy import Column, Integer, String, ForeignKey, Numeric, DateTime, Enum
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import enum

from app.database import Base


class OrderType(str, enum.Enum):
    BUY = "BUY"
    SELL = "SELL"


class OrderStatus(str, enum.Enum):
    PENDING = "PENDING"
    FILLED = "FILLED"
    CANCELLED = "CANCELLED"
    FAILED = "FAILED"


class VirtualAccount(Base):
    """虛擬交易帳戶"""
    __tablename__ = "virtual_accounts"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey('users.id', ondelete='CASCADE'), nullable=False, unique=True)

    # 資金
    initial_balance = Column(Numeric(15, 2), default=1000000)  # 初始資金 100萬
    cash_balance = Column(Numeric(15, 2), default=1000000)     # 現金餘額
    total_value = Column(Numeric(15, 2), default=1000000)      # 總資產價值

    # 損益
    total_profit_loss = Column(Numeric(15, 2), default=0)      # 總損益
    total_profit_loss_percent = Column(Numeric(6, 4), default=0)  # 總損益百分比

    # 時間戳
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # 關聯
    user = relationship("User", backref="virtual_account")
    positions = relationship("VirtualPosition", back_populates="account", cascade="all, delete-orphan")
    orders = relationship("VirtualOrder", back_populates="account", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<VirtualAccount user_id={self.user_id} balance={self.cash_balance}>"


class VirtualPosition(Base):
    """虛擬持倉"""
    __tablename__ = "virtual_positions"

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey('virtual_accounts.id', ondelete='CASCADE'), nullable=False)
    stock_id = Column(String(10), ForeignKey('stocks.stock_id'), nullable=False)

    # 持倉資訊
    quantity = Column(Integer, default=0)              # 持有股數
    avg_cost = Column(Numeric(10, 2))                  # 平均成本
    current_price = Column(Numeric(10, 2))             # 當前價格
    market_value = Column(Numeric(15, 2))              # 市值

    # 損益
    unrealized_pnl = Column(Numeric(15, 2), default=0)       # 未實現損益
    unrealized_pnl_percent = Column(Numeric(6, 4), default=0)  # 未實現損益百分比

    # 時間戳
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # 關聯
    account = relationship("VirtualAccount", back_populates="positions")
    stock = relationship("Stock")

    def __repr__(self):
        return f"<VirtualPosition {self.stock_id} qty={self.quantity}>"


class VirtualOrder(Base):
    """虛擬訂單"""
    __tablename__ = "virtual_orders"

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey('virtual_accounts.id', ondelete='CASCADE'), nullable=False)
    stock_id = Column(String(10), ForeignKey('stocks.stock_id'), nullable=False)

    # 訂單資訊
    order_type = Column(String(10), nullable=False)    # BUY, SELL
    quantity = Column(Integer, nullable=False)          # 委託數量
    price = Column(Numeric(10, 2), nullable=False)     # 委託價格
    filled_quantity = Column(Integer, default=0)       # 成交數量
    filled_price = Column(Numeric(10, 2))              # 成交價格

    # 狀態
    status = Column(String(20), default='PENDING')     # PENDING, FILLED, CANCELLED, FAILED

    # 時間戳
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    filled_at = Column(DateTime(timezone=True))

    # 關聯
    account = relationship("VirtualAccount", back_populates="orders")
    stock = relationship("Stock")

    def __repr__(self):
        return f"<VirtualOrder {self.order_type} {self.stock_id} qty={self.quantity}>"
